executor/src/core.rs

```diff
@@ -143,54 +147,175 @@ where
                 // the second batch since there is no point in executing twice the same transactions
                 // (as the second execution attempt will always fail).
                 debug!("Duplicate batch {batch_digest}");
-                self.execution_indices.skip_batch(total_batches);
-                return Ok(());
+                return Ok(Vec::new());
             }
         };
 
+        // The consensus simply orders bytes, so we first need to deserialize the transaction.
+        // If the deserialization fail it is safe to ignore the transaction since all correct
+        // clients will do the same. Remember that a bad authority or client may input random
+        // bytes to the consensus.
+        let transactions = transactions
+            .into_iter()
+            .map(
+                |serialized| match bincode::deserialize::<State::Transaction>(&serialized) {
```

Comment:
You mean to push the deserialization into the `ExecutorState` implementation itself?

Like I said I think letting the system know about the `Transaction` type is a good thing, it keeps your options open for hardening in the future by rejecting malicious input at the peripheries.

Reply:
I highlighted another change in the PR description, which is that `tx_output` will only contain the outcome for transactions that we managed to at least deserialize (previously it had raw bytes, now it's the model). It doesn't look like this would be a problem though, because you weren't able to tell the difference in the errors anyway, so you didn't now whether the raw bytes were legit without trying to deserialize again.

Let me know, though, because pushing deserialization into the state itself could restore full freedom.

Reply:
Moving it out would also put the decision of what serialization to use back with the application, ie. `bincode` would not be prescribed by this library. But you can achieve this by adding your own traits as well.

Reply:
Pushed a change which moves the deserialisation into the `SingleExecutor`, to restore the current behaviour of reporting the outcome of all transactions, even if they are malformed. The `BatchExecuttionState` now takes raw transactions, and it's up to the application to decide how much of it to process, and what deserialisation scheme to use.